app/src/main/java/org/mozilla/tv/firefox/MainActivity.kt

```diff
@@ -85,28 +54,22 @@ class MainActivity : LocaleAwareAppCompatActivity(), OnUrlEnteredListener, Media
         initMediaSession()
         lifecycle.addObserver(serviceLocator.webViewCache)
 
-        val intent = SafeIntent(intent)
-
         window.decorView.systemUiVisibility = View.SYSTEM_UI_FLAG_LAYOUT_STABLE or View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN
 
         setContentView(R.layout.activity_main)
 
-        IntentValidator.validateOnCreate(this, intent, savedInstanceState, ::onValidBrowserIntent)
+        val intent = SafeIntent(intent)
+        val intentData = IntentValidator.validateOnCreate(this, intent, savedInstanceState)
+
+        val session = getOrCreateSession(intentData)
 
-        webRenderComponents.sessionManager.register(sessionObserver, owner = this)
+        val screenController = serviceLocator.screenController
+        screenController.setUpFragmentsForNewSession(supportFragmentManager, session)
 
-        if (webRenderComponents.sessionManager.sessions.isEmpty()) {
-            serviceLocator.screenController.showBrowserScreenForUrl(
-                this@MainActivity,
-                supportFragmentManager,
-                URLs.APP_URL_HOME,
-                Session.Source.NONE
-            )
+        if (intentData != null) {
+            screenController.showBrowserScreenForUrl(supportFragmentManager, intentData.url)
```

Comment:
Good points - this startup code was very confusing (and still kind of is), and it would be helpful to be able clarify it into all the cases. We did document the cases that startup would go through but yes, it does bounce back and forth, due to how Android handles startup.
Trying to think if it would be better to file a follow-up for cleaning up the startup flow, or fix it here? I'm leaning towards a follow-up, so we can get this code landed.

Reply:
Follow-up in #1644